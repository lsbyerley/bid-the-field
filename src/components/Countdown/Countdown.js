import { useCountdown } from 'rooks';
import { secondsToHours, secondsToMinutes } from 'date-fns';

const Countdown = ({ auction = {}, setAuctionOver = () => {} }) => {
  const endTime = new Date(auction.end_date);

  const seconds = useCountdown(endTime, {
    interval: 1000,
    // onDown: (time) => console.log('onDown', time),
    onEnd: () => setAuctionOver(true),
  });

  if (seconds === 0) {
    // return null;
  }

  const hours = secondsToHours(seconds);
  const minutes = secondsToMinutes(seconds);

  const getHoursText = (hours) => {
    const hourText = hours > 1 ? 'hours' : 'hour';

    return (
      <span className='text-info'>
        ~{hours} {hourText}
      </span>
    );
  };

  const getMinutesText = (minutes) => {
    const minutesText = minutes > 1 ? 'minutes' : 'minute';
    return (
      <span className='text-success'>
        ~{minutes} {minutesText}
      </span>
    );
  };

  const getSecondsText = (seconds) => {
    const secondsText = seconds > 1 ? 'seconds' : 'second';
    return (
      <span className='text-warning'>
        {seconds === 0 && <>No time</>}
        {seconds > 0 && (
          <>
            {seconds} {secondsText}
          </>
        )}
      </span>
    );
  };

  return (
    <div className='mb-5 rounded-lg card card-compact bg-base-200'>
      <div className='items-center justify-center card-body'>
        <h3 className='text-lg font-medium '>
          {hours >= 1 && getHoursText(hours)}
          {minutes <= 59 && minutes >= 2 && getMinutesText(minutes)}
          {seconds <= 119 && getSecondsText(seconds)}
          <span className='text-sm'>&nbsp;remaining</span>
        </h3>
      </div>
    </div>
  );
};

export default Countdown;
