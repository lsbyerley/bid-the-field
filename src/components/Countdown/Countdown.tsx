import { useCountdown } from 'rooks';
import { secondsToHours, secondsToMinutes } from 'date-fns';

import type { CountdownProps } from '@/types';

const Countdown = ({
  auction,
  auctionStarted,
  setAuctionStarted = () => {},
  // auctionOver,
  setAuctionOver = () => {},
}: CountdownProps) => {
  const startTime = new Date(auction.start_date);
  const endTime = new Date(auction.end_date);

  const timeToUse = !auctionStarted ? startTime : endTime;
  const funcToCallOnEnd = !auctionStarted ? setAuctionStarted : setAuctionOver;

  const seconds = useCountdown(timeToUse, {
    interval: 1000,
    onEnd: () => funcToCallOnEnd(true),
  });

  const hours = secondsToHours(seconds);
  const minutes = secondsToMinutes(seconds);

  const getHoursText = (hours: number) => {
    const hourText = hours > 1 ? 'hours' : 'hour';

    return (
      <span className='text-info'>
        ~{hours} {hourText}
      </span>
    );
  };

  const getMinutesText = (minutes: number) => {
    const minutesText = minutes > 1 ? 'minutes' : 'minute';
    return (
      <span className='text-success'>
        ~{minutes} {minutesText}
      </span>
    );
  };

  const getSecondsText = (seconds: number) => {
    const secondsText = seconds > 1 ? 'seconds' : 'second';
    return (
      <span className='text-warning'>
        {seconds === 0 && <>No time</>}
        {seconds > 0 && (
          <>
            {seconds} {secondsText}
          </>
        )}
      </span>
    );
  };

  /*if (!auctionStarted) {
    return (
      <div className='rounded-lg card card-compact bg-base-200'>
        <div className='items-center justify-center card-body'>
          <h3 className='text-lg font-medium '>
            <span className='text-info'>not started</span>
          </h3>
        </div>
      </div>
    );
  }*/

  return (
    <div className='rounded-lg card card-compact bg-base-100'>
      <div className='items-center justify-center card-body'>
        <h3 className='text-lg font-medium '>
          {hours >= 1 && getHoursText(hours)}
          {minutes <= 59 && minutes >= 2 && getMinutesText(minutes)}
          {seconds <= 119 && getSecondsText(seconds)}
          {!auctionStarted && (
            <span className='text-sm'>&nbsp;until start</span>
          )}
          {auctionStarted && <span className='text-sm'>&nbsp;remaining</span>}
        </h3>
      </div>
    </div>
  );
};

export default Countdown;
