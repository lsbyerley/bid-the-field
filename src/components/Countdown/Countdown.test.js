import React from 'react';
import { render, screen } from 'test-utils';
import Countdown from './Countdown';

jest.mock('date-fns', () => ({
  format: jest.fn((props) => JSON.stringify(props)),
  secondsToHours: jest.fn((props) => JSON.stringify(props)),
  secondsToMinutes: jest.fn((props) => JSON.stringify(props)),
}));

/*jest.mock('rooks', () => ({
  useCountdown: jest.fn(),
}));*/

const defaultProps = (overrides) => {
  return {
    auction: {
      start_date: 'test',
      end_date: 'test',
    },
    auctionStarted: true,
    auctionOver: false,
    ...overrides,
  };
};

const renderCountdown = (props) => {
  const utils = render(<Countdown {...props} />);
  const home = utils.container.firstChild;
  return { ...utils, home };
};

describe('Countdown', () => {
  it('should render and match snapshot', () => {
    const { home } = renderCountdown(defaultProps());
    expect(home).toBeTruthy();
    expect(home).toMatchSnapshot();
  });
});
