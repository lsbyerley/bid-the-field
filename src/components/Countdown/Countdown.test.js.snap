// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown should render and match snapshot 1`] = `
<div
  class="rounded-lg card card-compact bg-base-100"
>
  <div
    class="items-center justify-center card-body"
  >
    <h3
      class="text-lg font-medium "
    >
      <span
        class="text-warning"
      >
        No time
      </span>
      <span
        class="text-sm"
      >
        Â remaining
      </span>
    </h3>
  </div>
</div>
`;
